### DISCUSSION TITLE
soft dependencies on core packages

### DISCUSSION THREAD

**Discussion by fmenezes** [Homebrew/discussions#3286] [Writing Formulae/Casks] [answered]
Hi there!

My use case basically is I want users of the [mongodb-atlas](https://github.com/Homebrew/homebrew-core/blob/1691f9cf173ccd402d2f7faae003ccea2b440abb/Aliases/mongodb-atlas) alias to install both [mongodb-atlas-cli](https://github.com/Homebrew/homebrew-core/blob/1691f9cf173ccd402d2f7faae003ccea2b440abb/Formula/mongodb-atlas-cli.rb) and [mongosh](https://github.com/Homebrew/homebrew-core/blob/1691f9cf173ccd402d2f7faae003ccea2b440abb/Formula/mongosh.rb). Ideally, this would be the case by default but optionally users would be able to install them separately.

I see in the [Formula Cookbook](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#specifying-other-formulae-as-dependencies) that we can add 4 types of dependencies `build`, `test`, `optional` and `recommended`.

I think for the use case I've mentioned flagging `mongosh` as a recommended dependency of `mongodb-atlas-cli` would achieve our goal. Unfortunately `optional` and `recommended` dependencies are not allowed in core packages.

Is there a way to achieve it in core packages?

**Comment by carlocab** [comment, marked as answer]
There is not, unfortunately. You can use optional or recommended dependencies if you host your formulae in an external tap.

**Reply by fmenezes** [reply]
thanks for the quick reply @carlocab

**Comment by fmenezes** [comment]
@carlocab how about having two packages one with the dependency and one without it? would it be a pattern the core team would accept?

**Reply by SMillerDev** [reply]
Does the software work without the dependency?

**Reply by fmenezes** [reply]
@SMillerDev yes it does, there are checks to see if the said dependency is installed or not.

**Reply by SMillerDev** [reply]
So the tool doesn't work because it'll complain about a missing dependency then?

**Reply by fmenezes** [reply]
it will work, optional features would be disabled

**Reply by fmenezes** [reply]
I'll be more specific `mongodb-atlas-cli` creates/manages the infrastructure but if you use `atlas quickstart` command it will create your cluster and also call `mongosh` to connect to it.

If `mongosh` is not installed it will create the cluster but it will not connect to it. Not to mention the vast majority of commands have no `mongosh` interaction.

**Reply by SMillerDev** [reply]
That seems reasonable enough to add as a permanent dependency. Duplicating the formula is really not a solution since maintainability for that would be terrible.

**Reply by fmenezes** [reply]
thanks for your answer @SMillerDev

**Comment by boooczek** [comment]
@SMillerDev how would you recommend handling the use case for CI / CD where users will need the mongodb-atlas-cli only as the size of the package with dependency will be much bigger and will impact their pipeline?
What do you think about a package that only contains let’s say a README that depends on mongodb-atlas-cli and mongosh?

**Reply by SMillerDev** [reply]
Make your own tap where you can define whatever dependencies you want. CI speed is not really a concern for Homebrew.
